### PR TITLE
Cli add functions

### PR DESCRIPTION
**Pull request by sayan-mitra**
**Fixes** # (*issue*)

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
## Checklist:
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added unit tests for the code
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: processor/stash/admin.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	DestName string
+}
+
+func (s *StashRpcHandler) getErrorCountByDest(dbHandle *sql.DB, jobTableName string) ([]DestinationCountResult, error) {
+	results := make([]DestinationCountResult, 0)
+	uniqueSourceValsStmt := fmt.Sprintf(`select count(*) as count, custom_val as dest from %s group by custom_val`, jobTableName)
+	var rows *sql.Rows
+	var err error
+	rows, err = dbHandle.Query(uniqueSourceValsStmt)
+	//defer rows.Close() // don't need to check error as we are firing read ops/ but for error above this will give error too, so calling at the end
```
Comment (SumanthPuram) [inline processor/stash/admin.go:45]: Did not understand this? Even for read, we need to close the rows right?

File: processor/stash/admin.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	}
+	singleResult := DestinationCountResult{}
+	for rows.Next() {
+		if err = rows.Scan(&singleResult.Count, &singleResult.DestName); err != nil {
+			return results, err
+		}
+		results = append(results, singleResult)
+	}
+
+	if err = rows.Err(); err != nil {
```
Comment (SumanthPuram) [inline processor/stash/admin.go:57]: Let's use the same pattern that we are using in the code?

File: router/router.go [hunk 3]
```diff
@@ -142,6 +147,7 @@ func loadConfig() {
 	retryTimeWindow = config.GetDuration("Router.retryTimeWindowInMins", time.Duration(180)) * time.Minute
 	minRetryBackoff = config.GetDuration("Router.minRetryBackoffInS", time.Duration(10)) * time.Second
 	maxRetryBackoff = config.GetDuration("Router.maxRetryBackoffInS", time.Duration(300)) * time.Second
+	maxFailedListCount = config.GetInt("Router.failedListCount", 10)
```
Comment (SumanthPuram) [inline router/router.go:150]: We use size more often, let's do the same.
`Router.failedEventsCacheSize`

File: router/router.go [hunk 4]
```diff
@@ -43,6 +44,9 @@ type HandleT struct {
 	perfStats                              *misc.PerfStats
 	successCount                           uint64
 	failCount                              uint64
+	failedListMutex                        sync.RWMutex
+	failedList                             *list.List
+	failedJobStatusChan                    chan jobsdb.JobStatusT
```
Comment (SumanthPuram) [inline router/router.go:49]: Let's call it failedEventsChan?

File: router/router.go [hunk 5]
```diff
@@ -43,6 +44,9 @@ type HandleT struct {
 	perfStats                              *misc.PerfStats
 	successCount                           uint64
 	failCount                              uint64
+	failedListMutex                        sync.RWMutex
+	failedList                             *list.List
```
Comment (SumanthPuram) [inline router/router.go:48]: Here also failedEventsList?

File: router/router.go [hunk 6]
```diff
@@ -583,6 +591,31 @@ func durationBeforeNextAttempt(attempt int) (d time.Duration) {
...
+	}
+	rt.failedList.PushFront(jobStatus)
+}
+
+func (rt *HandleT) readFailedJobStatusChan() {
+	for {
+		select {
+		case jobStatus := <-rt.failedJobStatusChan:
+			rt.addToFailedList(jobStatus)
+		case <-time.After(200):
```
Comment (SumanthPuram) [inline router/router.go:609]: Remove this?

**Review by SumanthPuram — COMMENTED** [review]
Cleanup needed

File: gateway/admin.go [hunk 7]
```diff
@@ -21,3 +31,147 @@ func (g *GatewayAdmin) Status() interface{} {
...
+
+	return sources
+}
+
+func (r *SqlRunner) getNumUniqueUsers() int {
+	var numUsers int
+	if r.err != nil {
+		return 0
+	}
+	numUniqueUsersStmt := fmt.Sprintf(`select count(*) from (select  distinct user_id from %s) as t`, r.jobTableName)
```
[the remark this answers is not in the document]
Comment (sayan-mitra) [reply gateway/admin.go:86]: I read somewhere the later is slower. will add the link

File: gateway/admin.go [hunk 8]
```diff
@@ -21,3 +31,147 @@ func (g *GatewayAdmin) Status() interface{} {
...
+// Average batch size ⇒ num_events we want per ds
+// writeKey, count(*)  we want source name to count per ds
+// Num Distinct users per ds
+// Avg Event size = Table_size / (avg Batch size * Total rows) is Table_size correct measure?
+func (g *GatewayRPCHandler) GetDSStats(dsName string, result *string) error {
+	jobTableName := prefix + dsName
+	dbHandle, err := sql.Open("postgres", g.jobsDB.GetConnectionStringPresent())
+	defer dbHandle.Close()
+	runner := &SqlRunner{dbHandle: dbHandle, jobTableName: jobTableName, err: err}
+	sources := runner.getUniqueSources()
```
[the remark this answers is not in the document]
Comment (sayan-mitra) [reply gateway/admin.go:143]: Modelled it as higher order error handling.
For any code path, if we encounter an error, we will immediately return for next calls without executing further.

File: gateway/gateway.go [hunk 9]
```diff
@@ -911,8 +912,10 @@ func (gateway *HandleT) backendConfigSubscriber() {
 		configSubscriberLock.Lock()
 		enabledWriteKeysSourceMap = map[string]string{}
 		enabledWriteKeyWebhookMap = map[string]string{}
+		enabledSourceIDToNameMap = map[string]string{}
 		sources := config.Data.(backendconfig.SourcesT)
 		for _, source := range sources.Sources {
+			enabledSourceIDToNameMap[source.ID] = source.Name
```
[the remark this answers is not in the document]
Comment (sayan-mitra) [reply gateway/gateway.go:918]: will change

File: jobsdb/jobsdb.go [hunk 10]
```diff
@@ -228,6 +230,10 @@ func (jd *HandleT) assert(cond bool, errorString string) {
 	}
 }
 
+func (jd *HandleT) GetConnectionStringPresent() string {
```
[the remark this answers is not in the document]
Comment (sayan-mitra) [reply jobsdb/jobsdb.go:233]: thought jobsdb should be responsible for supplying the connection string to all packages

File: jobsdb/jobsdb.go [hunk 11]
```diff
@@ -748,6 +754,7 @@ func (jd *HandleT) checkIfFullDS(ds dataSetT) bool {
 	}
 
 	return false
+
```
[the remark this answers is not in the document]
Comment (sayan-mitra) [reply jobsdb/jobsdb.go:757]: will remove

File: processor/processor.go [hunk 12]
```diff
@@ -250,6 +276,8 @@ func loadConfig() {
 	customDestinations = []string{"KAFKA", "KINESIS", "AZURE_EVENT_HUB"}
 	// EventSchemas feature. false by default
 	enableEventSchemasFeature = config.GetBool("EventSchemas.enableEventSchemasFeature", false)
+	maxItemsInTransformEventsByTimePQ = config.GetInt("Processor.maxItemsInTransformEventsByTimePQ", 5)
```
[the remark this answers is not in the document]
Comment (sayan-mitra) [reply processor/processor.go:279]: will change

File: processor/stash/admin.go (hunk 1 above)
Comment (sayan-mitra) [reply processor/stash/admin.go:45]: yes. We cannot defer close only for cases for ex: table is not present.
handled it separately below

File: gateway/admin.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (sayan-mitra) [reply gateway/admin.go:86]: https://stackoverflow.com/questions/11250253/postgresql-countdistinct-very-slow

File: gateway/admin.go [hunk 13]
```diff
@@ -21,3 +31,147 @@ func (g *GatewayAdmin) Status() interface{} {
...
+	if r.err = rows.Close(); r.err != nil {
+		return sources
+	}
+
+	return sources
+}
+
+func (r *SqlRunner) getNumUniqueUsers() int {
+	var numUsers int
+	if r.err != nil {
```
[the remark this answers is not in the document]
Comment (sayan-mitra) [reply gateway/admin.go:83]: returning from each function if we have encountered an error before
here opening db connection

File: gateway/admin.go [hunk 14]
```diff
@@ -21,3 +31,147 @@ func (g *GatewayAdmin) Status() interface{} {
...
+	return numUsers
+}
+
+func (r *SqlRunner) getAvgBatchSize() float64 {
+	var batchSize sql.NullFloat64
+	var avgBatchSize float64
+	if r.err != nil {
+		return 0
+	}
+	avgBatchSizeStmt := fmt.Sprintf(`select avg(jsonb_array_length(batch)) from (select event_payload->'batch' as batch from %s) t`, r.jobTableName)
```
[the remark this answers is not in the document]
Comment (sayan-mitra) [reply gateway/admin.go:97]: will check

File: processor/processor.go [hunk 15]
```diff
@@ -822,6 +850,9 @@ func (proc *HandleT) processJobsForDest(jobList []*jobsdb.JobT, parsedEventList
 
 	proc.destProcessing.Start()
 	proc.logger.Debug("[Processor: processJobsForDest] calling transformations")
+	logger.Debug("[Processor: processJobsForDest] calling transformations")
```
[the remark this answers is not in the document]
Comment (sayan-mitra) [reply processor/processor.go:853]: yes, will remove

File: gateway/admin.go [hunk 16]
```diff
@@ -21,3 +31,147 @@ func (g *GatewayAdmin) Status() interface{} {
...
+	}
+	sourceEvent := SourceEvents{}
+	for rows.Next() {
+		if r.err = rows.Scan(&sourceEvent.Count, &sourceEvent.Source); r.err != nil {
+			return sources
+		}
+		sources = append(sources, sourceEvent)
+	}
+
+	if r.err = rows.Err(); r.err != nil {
```
[the remark this answers is not in the document]
Comment (sayan-mitra) [reply gateway/admin.go:70]: done

File: gateway/admin.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (sayan-mitra) [reply gateway/admin.go:97]: it works but as with the above query, not sure if it is slow

File: gateway/gateway.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (sayan-mitra) [reply gateway/gateway.go:918]: done

File: jobsdb/jobsdb.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (sayan-mitra) [reply jobsdb/jobsdb.go:233]: removed, missed earlier that jobsdb package was already exporting this utility

File: jobsdb/jobsdb.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (sayan-mitra) [reply jobsdb/jobsdb.go:757]: done

File: processor/processor.go [hunk 17]
```diff
@@ -139,6 +145,22 @@ func NewProcessor() *HandleT {
 	}
 }
 
+func (proc *HandleT) Status() interface{} {
+	proc.transformEventsByTimeMutex.RLock()
+	defer proc.transformEventsByTimeMutex.RUnlock()
+	statusRes := make(map[string][]TransformRequestT)
+	for _, pqDestEvent := range proc.destTransformEventsByTimeTaken {
+		fmt.Println(pqDestEvent.ProcessingTime)
```
[the remark this answers is not in the document]
Comment (sayan-mitra) [reply processor/processor.go:153]: done

File: processor/processor.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (sayan-mitra) [reply processor/processor.go:279]: done

File: processor/processor.go [hunk 18]
```diff
@@ -851,8 +882,12 @@ func (proc *HandleT) processJobsForDest(jobList []*jobsdb.JobT, parsedEventList
 			} else {
 				// We need not worry about breaking up a single user sessions in this case
 				destStat.userTransform.Start()
+				beforeTransformRequest = time.Now()
```
[the remark this answers is not in the document]
Comment (sayan-mitra) [reply processor/processor.go:885]: changed names

File: processor/processor.go [hunk 19]
```diff
@@ -994,6 +1033,31 @@ func (proc *HandleT) processJobsForDest(jobList []*jobsdb.JobT, parsedEventList
...
+		return
+	}
+	pq.Add(event)
+}
+
+// Utility for tests
+func (proc *HandleT) printPQ() {
+	for {
+		select {
+		case <-time.After(1000000000):
```
[the remark this answers is not in the document]
Comment (sayan-mitra) [reply processor/processor.go:1054]: commented the tests

File: router/router.go (hunk 5 above)
Comment (sayan-mitra) [reply router/router.go:48]: done

File: router/router.go (hunk 3 above)
Comment (sayan-mitra) [reply router/router.go:150]: done

File: router/router.go (hunk 6 above)
Comment (sayan-mitra) [reply router/router.go:609]: removed

File: router/router.go (hunk 4 above)
Comment (sayan-mitra) [reply router/router.go:49]: done

File: gateway/admin.go [hunk 20]
```diff
@@ -21,3 +31,148 @@ func (g *GatewayAdmin) Status() interface{} {
...
+}
+
+type SourceEvents struct {
+	Count  int
+	Source string
+}
+
+func (r *SqlRunner) getUniqueSources() []SourceEvents {
+	sources := make([]SourceEvents, 0)
+	if r.err != nil {
```
Comment (SumanthPuram) [inline gateway/admin.go:52]: I don't think this is the right way of handling errors. Why can't we use the same standard across everything?

Comment (SumanthPuram) [reply gateway/admin.go:52]: Please move this up to the caller.

File: gateway/admin.go [hunk 21]
```diff
@@ -21,3 +31,148 @@ func (g *GatewayAdmin) Status() interface{} {
...
 }
+
+type GatewayRPCHandler struct {
+	jobsDB jobsdb.JobsDB
+}
+
+type SqlRunner struct {
+	dbHandle     *sql.DB
+	jobTableName string
+	err          error
```
Comment (SumanthPuram) [inline gateway/admin.go:42]: If we remove the err inside SqlRunner, this would make things compatible with what we already have.

**Review by SumanthPuram — CHANGES_REQUESTED** [review]

File: router/admin.go [hunk 22]
```diff
@@ -27,7 +35,217 @@ func (ra *RouterAdmin) Status() interface{} {
...
+	dsStats.err = rows.Err()
+	if dsStats.err != nil {
+		return
+	}
+	dsStats.err = rows.Close()
+	return
+}
+
+func getUnprocessedJobCounts(dbHandle *sql.DB, dsName string, dsStats *DSStats) {
+	routerTableName := "rt_jobs_" + dsName
```
Comment (SumanthPuram) [inline router/admin.go:245]: Add jobs in the variable name to make it easy to read?
Eg. `routerJobsTable, routerJobStatusTable`

File: router/admin.go [hunk 23]
```diff
@@ -27,7 +35,217 @@ func (ra *RouterAdmin) Status() interface{} {
...
+
+	dsStats.err = rows.Err()
+	if dsStats.err != nil {
+		return
+	}
+	dsStats.err = rows.Close()
+	return
+}
+
+func getJobStatusCountsByExecTime(dbHandle *sql.DB, dsName string, dsStats *DSStats) {
```
Comment (SumanthPuram) [inline router/admin.go:214]: Instead should we call this - `getLatestJobStatusCounts()`?

File: router/admin.go [hunk 24]
```diff
@@ -27,7 +35,217 @@ func (ra *RouterAdmin) Status() interface{} {
...
+type DSStats struct {
+	JobCountsByStateAndDestination           []JobCountsByStateAndDestination
+	FailedStatusErrorCodeCountsByDestination []FailedStatusErrorCodeCountsByDestination
+	JobCountByConnections                    []JobCountByConnections
+	JobStatusCountsByExecTime                []JobStatusCountsByExecTime
+	UnprocessedJobCounts                     int
+	err                                      error `json:"-"`
+}
+
+// group_by job_status
```
Comment (SumanthPuram) [inline router/admin.go:90]: Let's add some example outputs here? To be able to understand easily what each function does.

File: gateway/admin.go (hunk 20 above)
Comment (sayan-mitra) [reply gateway/admin.go]: done

File: gateway/admin.go (hunk 21 above)
Comment (sayan-mitra) [reply gateway/admin.go]: done

File: processor/processor.go [hunk 25]
```diff
@@ -1004,6 +1036,20 @@ func getTruncatedEventList(jobList []*jobsdb.JobT, maxEvents int) (truncatedList
 	return jobList, totalEvents
 }
 
+func (proc *HandleT) addToTransformEventByTimePQ(event *TransformRequestT, pq *transformRequestPQ) {
+	proc.transformEventsByTimeMutex.Lock()
+	defer proc.transformEventsByTimeMutex.Unlock()
+	if pq.Len() == transformTimesPQLength {
```
[the remark this answers is not in the document]
Comment (sayan-mitra) [reply processor/processor.go]: done

File: router/admin.go (hunk 23 above)
Comment (sayan-mitra) [reply router/admin.go]: done

File: router/admin.go (hunk 24 above)
Comment (sayan-mitra) [reply router/admin.go:91]: added table format

File: gateway/admin.go [hunk 26]
```diff
@@ -21,3 +32,170 @@ func (g *GatewayAdmin) Status() interface{} {
...
+		}
+		sources = append(sources, sourceEvent)
+	}
+
+	err = rows.Err()
+	if err != nil {
+		return sources, err
+	}
+
+	err = rows.Close() // rows.close will be called in defer too, but it should be harmless to call multiple times
```
Comment (SumanthPuram) [inline gateway/admin.go:74]: Remove this.

File: gateway/admin.go [hunk 27]
```diff
@@ -21,3 +32,170 @@ func (g *GatewayAdmin) Status() interface{} {
...
+	sourceEvent := SourceEvents{}
+	for rows.Next() {
+		err = rows.Scan(&sourceEvent.Count, &sourceEvent.Source)
+		if err != nil {
+			return sources, err // defer closing of rows, so return will not memory leak
+		}
+		sources = append(sources, sourceEvent)
+	}
+
+	err = rows.Err()
```
Comment (SumanthPuram) [inline gateway/admin.go:69]: Follow similar pattern with what we are doing.

File: gateway/admin.go [hunk 28]
```diff
@@ -21,3 +32,170 @@ func (g *GatewayAdmin) Status() interface{} {
...
+	sources := make([]SourceEvents, 0)
+	uniqueSourceValsStmt := fmt.Sprintf(`select count(*) as count, parameters -> 'source_id' as source from %s group by parameters -> 'source_id';`, r.jobTableName)
+	rows, err = r.dbHandle.Query(uniqueSourceValsStmt)
+	if err != nil {
+		return sources, err
+	}
+	defer rows.Close()
+	sourceEvent := SourceEvents{}
+	for rows.Next() {
+		err = rows.Scan(&sourceEvent.Count, &sourceEvent.Source)
```
Comment (SumanthPuram) [inline gateway/admin.go:62]: We should display source name and source ID.

File: gateway/admin.go [hunk 29]
```diff
@@ -21,3 +32,170 @@ func (g *GatewayAdmin) Status() interface{} {
...
+func (g *GatewayRPCHandler) GetDSStats(dsName string, result *string) error {
+	var completeErr error
+	jobTableName := prefix + dsName
+	dbHandle, err := sql.Open("postgres", jobsdb.GetConnectionString())
+	if err != nil {
+		return err
+	}
+	defer dbHandle.Close() // since this also returns an error, we can explicitly close but not doing
+	runner := &SqlRunner{dbHandle: dbHandle, jobTableName: jobTableName}
+	sources, serr := runner.getUniqueSources()
```
Comment (SumanthPuram) [inline gateway/admin.go:151]: Use standard naming? Maybe err for all.

File: gateway/admin.go [hunk 30]
```diff
@@ -21,3 +32,170 @@ func (g *GatewayAdmin) Status() interface{} {
...
+	configSubscriberLock.RUnlock()
+	response, err := json.MarshalIndent(DSStats{sourcesEventToCounts, numUsers, avgBatchSize, tableSize, numRows}, "", " ")
+	if err != nil {
+		*result = ""
+		misc.AppendError("MarshalIndent", &completeErr, &err)
+	} else {
+		*result = string(response)
+	}
+
+	return err
```
Comment (SumanthPuram) [inline gateway/admin.go:189]: Return completeErr?

File: gateway/admin.go [hunk 31]
```diff
@@ -21,3 +32,170 @@ func (g *GatewayAdmin) Status() interface{} {
...
+}
+
+type SqlRunner struct {
+	dbHandle     *sql.DB
+	jobTableName string
+}
+
+type SourceEvents struct {
+	Count  int
+	Source string
```
Comment (SumanthPuram) [inline gateway/admin.go:47]: Rename this to name and also add ID.

File: gateway/admin.go [hunk 32]
```diff
@@ -21,3 +32,170 @@ func (g *GatewayAdmin) Status() interface{} {
...
+
+	return err
+}
+
+func runSQL(runner *SqlRunner, query string, reciever interface{}) error {
+	row := runner.dbHandle.QueryRow(query)
+	err := row.Scan(reciever)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			fmt.Println("Zero rows found")
```
Comment (SumanthPuram) [inline gateway/admin.go:197]: Either return nil here or don't handle this at all.

File: processor/priorityqueue.go [hunk 33]
```diff
@@ -66,3 +69,72 @@ func (pq *pqT) Print() {
 		pkgLogger.Debug(*v)
 	}
 }
+
+type TransformRequestT struct {
```
Comment (SumanthPuram) [inline processor/priorityqueue.go:73]: Can we remove the dependency with this struct in this file ? Not needed, but if there is a better way, we should do it.

File: processor/stash/admin.go [hunk 34]
```diff
@@ -0,0 +1,78 @@
...
+│───────│─────────────│
+*/
+func (s *StashRpcHandler) GetDSStats(dsName string, result *string) error {
+	jobTableName := prefix + dsName
+	dbHandle, err := sql.Open("postgres", jobsdb.GetConnectionString())
+	defer dbHandle.Close()
+	if err != nil {
+		return err
+	}
+	results, execErr := s.getErrorCountByDest(dbHandle, jobTableName)
```
Comment (SumanthPuram) [inline processor/stash/admin.go:33]: Group by error too.

File: processor/stash/admin.go [hunk 35]
```diff
@@ -0,0 +1,78 @@
...
+		}
+		results = append(results, singleResult)
+	}
+
+	err = rows.Err()
+	if err != nil {
+		return results, err
+	}
+
+	err = rows.Close()
```
Comment (SumanthPuram) [inline processor/stash/admin.go:72]: Remove extra close and err.

File: router/admin.go [hunk 36]
```diff
@@ -27,7 +37,285 @@ func (ra *RouterAdmin) Status() interface{} {
...
+	result := FailedStatusErrorCodeCountsByDestination{}
+	for rows.Next() {
+		err = rows.Scan(&result.Count, &result.ErrorCode, &result.Destination)
+		if err != nil {
+			return err
+		}
+		dsStats.FailedStatusErrorCodeCountsByDestination = append(dsStats.FailedStatusErrorCodeCountsByDestination, result)
+	}
+
+	err = rows.Err()
```
Comment (SumanthPuram) [inline router/admin.go:222]: Remove this too.

**Review by SumanthPuram — CHANGES_REQUESTED** [review]

File: processor/stash/admin.go (hunk 34 above)
Comment (bipin-rudder) [reply processor/stash/admin.go:33]: done

File: processor/stash/admin.go (hunk 2 above)
Comment (bipin-rudder) [reply processor/stash/admin.go:68]: done

File: gateway/admin.go (hunk 28 above)
Comment (bipin-rudder) [reply gateway/admin.go]: done

File: gateway/admin.go (hunk 27 above)
Comment (bipin-rudder) [reply gateway/admin.go]: done

File: gateway/admin.go (hunk 26 above)
Comment (bipin-rudder) [reply gateway/admin.go]: done

File: gateway/admin.go (hunk 29 above)
Comment (bipin-rudder) [reply gateway/admin.go]: done

File: gateway/admin.go (hunk 30 above)
Comment (bipin-rudder) [reply gateway/admin.go]: done

File: gateway/admin.go (hunk 31 above)
Comment (bipin-rudder) [reply gateway/admin.go]: done

File: gateway/admin.go (hunk 32 above)
Comment (bipin-rudder) [reply gateway/admin.go]: done

File: processor/stash/admin.go (hunk 35 above)
Comment (bipin-rudder) [reply processor/stash/admin.go]: done

File: router/admin.go (hunk 36 above)
Comment (bipin-rudder) [reply router/admin.go]: done

File: router/admin.go (hunk 22 above)
Comment (bipin-rudder) [reply router/admin.go]: done

File: processor/stash/admin.go (hunk 34 above)
Comment (chandumlg) [reply processor/stash/admin.go:33]: handle execErr

File: router/admin.go [hunk 37]
```diff
@@ -27,7 +37,285 @@ func (ra *RouterAdmin) Status() interface{} {
...
+	UnprocessedJobCounts           int
+}
+
+// group_by job_status
+// group by custom_val
+// Get all errors = distinct (error), count(*) where state=failed
+// Distinct (src_id, dst_id)
+// Router jobs status flow ⇒ ordered by rank
+// unprocessed_params ⇒ Num jobs not yet picked
+func (r *RouterRpcHandler) GetDSStats(dsName string, result *string) error {
```
Comment (chandumlg) [inline router/admin.go:98]: Handle panics in recover

File: gateway/admin.go [hunk 38]
```diff
@@ -21,3 +32,169 @@ func (g *GatewayAdmin) Status() interface{} {
...
+	return numUsers, err
+}
+
+func (r *SqlRunner) getAvgBatchSize() (float64, error) {
+	var batchSize sql.NullFloat64
+	var avgBatchSize float64
+	var err error
+	avgBatchSizeStmt := fmt.Sprintf(`select avg(jsonb_array_length(batch)) from (select event_payload->'batch' as batch from %s) t`, r.jobTableName)
+	err = runSQL(r, avgBatchSizeStmt, &batchSize)
+	if batchSize.Valid {
```
Comment (chandumlg) [inline gateway/admin.go:91]: Handle err. This can cause nil pointer exception;

File: gateway/admin.go [hunk 39]
```diff
@@ -21,3 +32,169 @@ func (g *GatewayAdmin) Status() interface{} {
...
+// Avg Event size = Table_size / (avg Batch size * Total rows) is Table_size correct measure?
+// add job status group by
+/*
+EventsBySource
+================================================================================
+│───────│────────│
+│ COUNT │ SOURCE │
+│───────│────────│
+│   448 │ tr-2   │
+│───────│────────│
```
Comment (chandumlg) [inline gateway/admin.go:134]: Update the table

File: router/admin.go [hunk 40]
```diff
@@ -27,7 +37,285 @@ func (ra *RouterAdmin) Status() interface{} {
...
+│ COUNT │ SOURCEID                      │ DESTINATIONID                 │
+│───────│───────────────────────────────│───────────────────────────────│
+│   323 │ "1kXnQTrRjEmjU2wH8KjRR8EJ3gm" │ "1kXo508bX4OAynyYkEBpH6aQYHP" │
+│   323 │ "1kXnQTrRjEmjU2wH8KjRR8EJ3gm" │ "1kYW7q5ApiMkIG9TGsSZb7PIlrf" │
+│   323 │ "1kXnQTrRjEmjU2wH8KjRR8EJ3gm" │ "1kgadfXiXiZPM8oKAtkPFxFjm0P" │
+│───────│───────────────────────────────│───────────────────────────────│
+*/
+func getJobCountByConnections(dbHandle *sql.DB, dsName string, dsStats *DSStats) error {
+	routerJobsTableName = "rt_jobs_" + dsName
+	sqlStmt := fmt.Sprintf(`select count(*), parameters->'source_id' as s, parameters -> 'destination_id' as d
```
Comment (chandumlg) [inline router/admin.go:242]: indent

File: router/admin.go [hunk 41]
```diff
@@ -27,7 +37,285 @@ func (ra *RouterAdmin) Status() interface{} {
...
+│         336 │ failed    │ 5    │
+│         336 │ executing │ 6    │
+│          68 │ waiting   │ 7    │
+│          68 │ executing │ 8    │
+│─────────────│───────────│──────│
+*/
+func getLatestJobStatusCounts(dbHandle *sql.DB, dsName string, dsStats *DSStats) error {
+	routerJobStatusTableName = "rt_job_status_" + dsName
+	sqlStmt := fmt.Sprintf(`SELECT COUNT(*), job_state, rank FROM (SELECT job_state, RANK() OVER(PARTITION BY job_id
+							ORDER BY exec_time DESC) as rank, job_id from %s) as inner_table GROUP BY rank, job_state order by rank, job_state`, routerJobStatusTableName)
```
Comment (chandumlg) [inline router/admin.go:289]: indent

File: gateway/admin.go (hunk 38 above)
Comment (bipin-rudder) [reply gateway/admin.go:91]: Just checked, batchSize.Valid will be true only when batchSize is NOT NULL, false when NULL. Hence wont cause a nil pointer exception

File: router/admin.go (hunk 40 above)
Comment (bipin-rudder) [reply router/admin.go]: done

File: router/admin.go (hunk 41 above)
Comment (bipin-rudder) [reply router/admin.go]: done

File: gateway/admin.go (hunk 39 above)
Comment (bipin-rudder) [reply gateway/admin.go]: done

File: router/admin.go (hunk 37 above)
Comment (bipin-rudder) [reply router/admin.go]: done

**Comment by bipin-rudder** [comment]
rebasing became difficult as the number of commits were high!
Squashed and raised a new PR for the same : https://github.com/rudderlabs/rudder-server/pull/748
This PR will be abandoned after its alternative gets merged!

**Pull request closed without merging**